File: Programmers/12973.cpp
#include <iostream>
#include <string>
#include <stack>

using namespace std;

int solution(string s) {
    stack<char> stk;
    int i = s.size() - 1;
    while(i >= 0) {
        if (!stk.empty() && stk.top() == s[i])
            stk.pop(); 
        else
            stk.push(s[i]);
        i--;
    }
    
    if(stk.empty()) return 1;
    else return 0;
}

// 시간초과 나옴. 정확성은 통과.. 스택 이용해야 한다고 함..
// int solution(string s) {
//     // vector<string> v = {"aa", "bb", "cc", "dd", "ee", "ff", "gg", "hh", "ii", "jj", "kk", "ll", "mm", "nn", "oo", "pp", "qq", "rr", "ss", "tt", "uu", "vv", "ww", "xx", "yy", "zz"};
   
//     int i = 0;
//     while(true) {    
        
//         if(s.empty() || i >= s.length()) {
//             break;
//         } 
//         if(s[i] == s[i + 1]) {
//             cout << s << endl;
//             s.erase(i, 2);
//             i = -1;
//         }
//         i++;
//     }
    
//     if (s.length() == 0) {
//         return 1;
//     } else {
//         return 0;
//     }
    
//     return 0;
// }

int main() {
    cout << solution("baabaa") << endl;
    return 0;
}




// int solution(string s)
// {
//     int answer = -1;
//     int i  = 1;
//     string temp = s;
//     int j = 1;
//     while(true) {
//         if (temp.length() == 0) {
//             answer = 1;
//             break;
//         }
//         else {
//             if(temp[i - 1] == temp[i]) {
//                 cout << "temp[i-1] == temp[i] : " << temp[i-1] << temp[i] << endl;
//                 string front = temp.substr(0, i - 1);
//                 string back = temp.substr(i+1, temp.length() - (i + 1));
//                 cout << "front: " << front << ", back: " << back << endl;
//                 temp = front + back;
//                 cout << "temp : " << temp << endl;
//                 i -= 2;
//                 j += 2;
//             } else if (j == s.length()) {
//                 answer = 0;
//                 break;
//             }
//             i++;
//             j++;
//         }
        
//     }

//     return answer;
// }

File: Python/programmers2-3/12973.py
def solution(s):
    
    stack = []
    
    for str in s:
        if len(stack) == 0:
            stack.append(str)
        else:
            if stack[-1] == str:
                stack.pop()
            else:
                stack.append(str)
                
    if len(stack) > 0:
        return 0
    else:
        return 1
            
    
    
    '''
    answer = -1
    temp = s
    same_idx = 0
    is_same = False
    while True:
        for idx in range(1, len(temp)):
            if temp[idx-1] == temp[idx]:
                is_same = True
                same_idx = idx

        if is_same == True:
            temp = temp[0:same_idx-1] + temp[same_idx+1:]
        else:
            break
        is_same = False
        same_idx = 0
    if len(temp) == 0:
        return 1
    else:
        return 0
    '''
    